balancer/pickfirst/pickfirst.go

```diff
@@ -103,22 +350,99 @@ func (b *pickfirstBalancer) ResolverError(err error) {
 	})
 }
 
+func (b *pickfirstBalancer) unsetSelectedSubConn() {
+	if b.selectedSubConn != nil {
+		b.selectedSubConn.subConn.Shutdown()
+		b.selectedSubConn = nil
+	}
+}
+
+func (b *pickfirstBalancer) goIdle() {
+	b.unsetSelectedSubConn()
+	b.refreshSunConnList()
+	if len(b.subConnList.subConns) == 0 {
+		b.state = connectivity.TransientFailure
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.TransientFailure,
+			Picker:            &picker{err: fmt.Errorf("empty address list")},
+		})
+		b.unsetSelectedSubConn()
+		b.subConnList.close()
+		b.cc.ResolveNow(resolver.ResolveNowOptions{})
+		return
+	}
+
+	callback := func() {
+		b.subConnList.startConnectingIfNeeded()
+	}
+
+	nextState := connectivity.Idle
+	if b.state == connectivity.TransientFailure {
+		// We stay in TransientFailure until we are Ready. See A62.
+		nextState = connectivity.TransientFailure
+	} else {
+		b.state = connectivity.Idle
+	}
+	b.cc.UpdateState(balancer.State{
+		ConnectivityState: nextState,
+		Picker: &idlePicker{
+			callback: callback,
+		},
+	})
+}
+
 type Shuffler interface {
 	ShuffleAddressListForTesting(n int, swap func(i, j int))
 }
 
 func ShuffleAddressListForTesting(n int, swap func(i, j int)) { rand.Shuffle(n, swap) }
 
+func (b *pickfirstBalancer) refreshSunConnList() {
+	subConnList := newSubConnList(b.latestAddressList, b)
+
+	// Reset the previous subConnList to release resources.
+	if b.subConnList != nil {
+		if b.logger.V(2) {
+			b.logger.Infof("Closing older subConnList")
+		}
+		b.subConnList.close()
+	}
+	b.subConnList = subConnList
+}
+
+func (b *pickfirstBalancer) connectUsingLatestAddrs() error {
+	b.refreshSunConnList()
+	if len(b.subConnList.subConns) == 0 {
+		b.state = connectivity.TransientFailure
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.TransientFailure,
+			Picker:            &picker{err: fmt.Errorf("empty address list")},
+		})
+		b.unsetSelectedSubConn()
+		b.subConnList.close()
+		return balancer.ErrBadResolverState
+	}
```

Comment:
I see this as an error handling step that should have happened when creating the subConnList ran into errors.

Reply:
Moved the error handling logic into the `refreshSubConnList` method. I can't move it into the `newSubConnList` method since its used to create an empty subConnList in the builder where is it expected for the list to be empty initially.